tool/lib/commands/latest_flutter_candidate.dart

```diff
@@ -0,0 +1,277 @@
+        '${auth}api.github.com',
+        '/repos/flutter/flutter/branches',
+        {
+          'per_page': '$maxPerPage',
+          'page': '$requestPage',
+        },
+      );
+
+      final List<Map<String, dynamic>> branches =
+          (jsonDecode((await http.get(uri)).body) as List)
+              .cast<Map<String, dynamic>>();
+
+      final branchNames =
+          branches.map((branch) => branch['name']).cast<String>();
+      allBranchNames.addAll(branchNames);
+
+      final candidateBranchesNames =
+          branchNames.where((name) => name.contains('candidate'));
+
+      for (final branchName in candidateBranchesNames) {
+        final semVer = semVerFromCandidateBranch(branchName);
+        if (semVer != null && semVer > latest) {
+          latest = semVer;
+          latestBranchName = branchName;
+        }
+      }
+
+      requestPage++;
+      if (branches.length < maxPerPage) {
+        lastPageReceived = true;
+      }
+    }
+
+    if (latestBranchName == null) {
+      print('Something went wrong. Could not find the latest candidate branch:'
+          '${allBranchNames.join('\n')}');
+    } else {
+      print(latestBranchName);
+    }
+  }
+}
+
+SemanticVersion? semVerFromCandidateBranch(String branch) {
+  final candidateBranchPattern =
+      RegExp(r'^flutter-([0-9]+).([0-9]+)-candidate.([0-9]+)$');
+
+  final match = candidateBranchPattern.firstMatch(branch);
+  if (match == null) {
+    return null;
+  }
+
+  final x = int.parse(match.group(1)!);
+  final y = int.parse(match.group(2)!);
+  final z = int.parse(match.group(3)!);
+
+  return SemanticVersion(major: x, minor: y, patch: z);
+}
+
+// TODO(kenz): consider placing this class in devtools_shared, as this is
+// currently copied from `devtools_app/lib/src/shared/version.dart`
+class SemanticVersion with CompareMixin {
+  SemanticVersion({
+    this.major = 0,
+    this.minor = 0,
+    this.patch = 0,
+    this.preReleaseMajor,
+    this.preReleaseMinor,
+  });
+
+  factory SemanticVersion.parse(String? versionString) {
+    if (versionString == null) return SemanticVersion();
+
+    // Remove any build metadata, denoted by a '+' character and whatever
+    // follows.
+    final buildMetadataIndex = versionString.indexOf('+');
+    if (buildMetadataIndex != -1) {
+      versionString = versionString.substring(0, buildMetadataIndex);
+    }
+
+    // [versionString] is expected to be of the form for VM.version, Dart, and
+    // Flutter, respectively:
+    // 2.15.0-233.0.dev (dev) (Mon Oct 18 14:06:26 2021 -0700) on "ios_x64"
+    // 2.15.0-178.1.beta
+    // 2.6.0-12.0.pre.443
+    //
+    // So split on the spaces to the version, and then on the dash char to
+    // separate the main semantic version from the pre release version.
+    final splitOnSpaces = versionString.split(' ');
+    final version = splitOnSpaces.first;
+    final splitOnDash = version.split('-');
+    assert(splitOnDash.length <= 2, 'version: $version');
```

Comment:
it might be way more simple to parse this using a regex.
https://github.com/flutter/devtools/blob/5ccb2bd3594456efd37a36b58ea793695875c567/tool/update_version.dart#L63-L70
is a good example of how to pull things apart.
I believe the regex for a candidate branch would be something like:
final semVerMatch = RegExp(r'^flutter-(?<major>\d+)\.(?<minor>\d+)-candidate\.(?<candidate_minor>\d+)')
      .firstMatch(version);

Reply:
It looks like i missunderstood this. I thought this was trying to parse candidate branches based on context. I see that is done elsewhere now though

Reply:
that being said this could definitly benefit from some regex named grouping. right now it is a bit hard to make sense of :sweat_smile:

Reply:
I'm going to hold off on this for now. This code was copied from the devtools_app repo (I have a comment above to consider placing this in devtools_shared), but any fixes to the semantic version code should be applied to the original file (or a shared location in the future).